### PR TITLE
(WIP) coord: refactor catalog content migrations

### PR DESCRIPTION
**Pull request by sploiselle**
@benesch Need to replace docs and figure out the semantics for determining versions, etc., but how does this general approach look?

### CODE REVIEW

**Comment by sploiselle** [comment]
The testdrive test failures seem to come from testdrive itself [opening the catalog](https://github.com/MaterializeInc/materialize/blob/f656bd6331a88ad2261ef3a9d4d75276d21ce86a/src/testdrive/src/action/sql.rs#L170-L172), i.e. I am able to get it to pass all of the tests through testdrive by ensuring the linked-to lines don't execute.

**Comment by benesch** [comment]
I can help you fix that tomorrow! We need to make sure that testdrive’s
catalog checking doesn’t actually write to the db.

This broadly looks great! Tsym! A few thoughts:

It occurs to me that the confluent_wire_protocol migration isn’t a great
fit for this new framework. That thing needs to run only once, to avoid
incorrectly flipping the default on Avro sources created post v0.7.1. I
think.

We should run the entire catalog migration process in a txn, and only
commit it once the catalog has successfully loaded. Assuming that can be
made to work. That way if we can’t read your catalog, at least we won’t
brick your ability to use the catalog with your previous version of
Materialize.

**Review by benesch — COMMENTED** [review]
Oops, forgot to press submit on this.

File: test/catalog-compat/catcompatck/Dockerfile [hunk 1]
```diff
@@ -27,4 +25,7 @@ COPY --from=edge /usr/local/bin/materialized /usr/local/bin/materialized-edge
 COPY --from=testdrive /usr/local/bin/testdrive /usr/local/bin/testdrive
 COPY catcompatck /usr/local/bin/catcompatck
 
+# Allow running this test with dev builds
+ENV MZ_DEV=1
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply test/catalog-compat/catcompatck/Dockerfile:29]: Have this fixed locally and will include it in my next push to this PR.

**Comment by sploiselle** [comment]
This is now ready to more fully review.

@benesch We chatted about setting the default value for all Avro schemas to `confluent_wire_format=false`, but that caused a panic in at least 1 Avro test where the records are packed into lists. Setting the value to `true` doesn't have the same issue, though. So the commit `sql: default confluent_wire_format=true for all new Avro schemas` is literally the opposite of what we agreed on, but works with the existing tests. Not sure if this is actually fine, or if we should wait for @quodlibetor to return to determine what the correct solution is. I'm also happy to dig into it, but it felt off the beaten path from what I set out to do.

File: src/build-info/src/lib.rs [hunk 2]
```diff
@@ -39,7 +39,11 @@ pub const DUMMY_BUILD_INFO: BuildInfo = BuildInfo {
 impl BuildInfo {
     /// Constructs a human-readable version string.
     pub fn human_version(&self) -> String {
-        format!("v{} ({})", self.version, &self.sha[..9])
+        if self == &DUMMY_BUILD_INFO {
```
Comment (benesch) [inline src/build-info/src/lib.rs:42]: Whoop, this seems like treating the symptom but not the cause. I think we should just come up with a better `DUMMY_BUILD_INFO`. Maybe stick "v0.0.0+dummy" in the `version` field and `0000000000000000000000000000000000000000` in the `she` field?

File: src/build-info/src/lib.rs [hunk 3]
```diff
@@ -30,16 +30,20 @@ pub struct BuildInfo {
...
     target_triple: "",
 };
 
 impl BuildInfo {
     /// Constructs a human-readable version string.
     pub fn human_version(&self) -> String {
-        format!("v{} ({})", self.version, &self.sha[..9])
+        if self == &DUMMY_BUILD_INFO {
+            "DUMMY_BUILD_INFO".to_string()
+        } else {
```
Comment (benesch) [inline src/build-info/src/lib.rs:44]: The new contents of `DUMMY_BUILD_INFO` LGTM but we should be able to ditch the `if` now!

File: src/coord/src/catalog/storage.rs [hunk 4]
```diff
@@ -277,36 +275,38 @@ impl Connection {
...
+            },
             None => {
                 tx.execute(
-                    "INSERT INTO settings (name, value) VALUES ('catalog_content_version', 0);",
+                    "INSERT INTO settings (name, value) VALUES ('catalog_content_version', 'new');",
                     params![],
                 )?;
-                0
+                "new".to_string()
             }
```
Comment (benesch) [inline src/coord/src/catalog/storage.rs:299]: Oh this is kinda weird now. There is the possibility of writing `new` into the database like this. I feel like we should ditch this branch and then use an `UPSERT` down in `set_catalog_content_version` below?

**Review by benesch — COMMENTED** [review]
This looks really good! We're still sketchy with our transactions at start up; I'm going to take a stab at this. (I'm just little leery of sending you down the path without trying it first. Gimme just a few!)

File: src/build-info/src/lib.rs (hunk 3 above)
Comment (sploiselle) [reply src/build-info/src/lib.rs]: :face_palm:

**Review by benesch — APPROVED** [review]
Scratch my idea. I checked out the code and this seems totally great for now. Thank you so much for fighting this fight!

**Comment by benesch** [comment]
Ugh, I think the migration has been safe this whole time, actually. I pushed up a commit that restores it with justification, based on your earlier revision that included it. 🙈

**Pull request merged**